source/extensions/tracers/opentelemetry/span_context_extractor.cc

```diff
@@ -0,0 +1,98 @@
+#include "span_context_extractor.h"
+
+#include "envoy/common/exception.h"
+#include "envoy/tracing/http_tracer.h"
+
+#include "source/common/http/header_map_impl.h"
+
+#include "absl/strings/escaping.h"
+#include "span_context.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Tracers {
+namespace OpenTelemetry {
+namespace {
+
+const Http::LowerCaseString& openTelemetryPropagationHeader() {
+  CONSTRUCT_ON_FIRST_USE(Http::LowerCaseString, "traceparent");
+}
+
+// See https://www.w3.org/TR/trace-context/#traceparent-header
+constexpr int kTraceparentHeaderSize = 55; // 2 + 1 + 32 + 1 + 16 + 1 + 2
+constexpr int kVersionHexSize = 2;
+constexpr int kTraceIdHexSize = 32;
+constexpr int kParentIdHexSize = 16;
+constexpr int kTraceFlagsHexSize = 2;
+
+bool isValidHex(const std::string& input) {
+  return std::all_of(input.begin(), input.end(),
+                     [](const char& c) { return absl::ascii_isxdigit(c); });
+}
+
+bool isAllZeros(const std::string& input) {
+  return std::all_of(input.begin(), input.end(), [](const char& c) { return c == '0'; });
+}
+
+} // namespace
+
+SpanContextExtractor::SpanContextExtractor(Tracing::TraceContext& trace_context)
+    : trace_context_(trace_context) {}
+
+SpanContextExtractor::~SpanContextExtractor() = default;
+
+bool SpanContextExtractor::propagationHeaderPresent() {
+  auto propagation_header = trace_context_.getByKey(openTelemetryPropagationHeader());
+  return propagation_header.has_value();
+}
+
+SpanContext SpanContextExtractor::extractSpanContext() {
+  auto propagation_header = trace_context_.getByKey(openTelemetryPropagationHeader());
+  if (!propagation_header.has_value()) {
+    // We should have already caught this, but just in case.
+    throw ExtractorException("No propagation header found");
+  }
```

Comment:
In general, no exceptions in data path?

Reply:
+1, you can use the `absl::StatusOr` for this case.

Reply:
Ah great call, updated this and in the usage/tests. Thanks!